fw/preamp/src/fans.c

```diff
@@ -0,0 +1,237 @@
+ *
+ * Returns the desired fan percent in Q7.8 fixed-point format.
+ */
+int16_t fanPercentFromTemps(int16_t amp_temp, int16_t psu_temp,
+                            int16_t rpi_temp) {
+  // Calculate fan percent for each temp
+  // measurement in Q7.8 format, 1.0 = 100%
+  int16_t amp_pcnt = (amp_temp - TEMP_AMP_THRESH_LOW_Q7_8) /
+                     (TEMP_AMP_THRESH_HIGH_C - TEMP_AMP_THRESH_LOW_C);
+  int16_t psu_pcnt = (psu_temp - TEMP_PSU_THRESH_LOW_Q7_8) /
+                     (TEMP_PSU_THRESH_HIGH_C - TEMP_PSU_THRESH_LOW_C);
+  int16_t rpi_pcnt = (rpi_temp - TEMP_RPI_THRESH_LOW_Q7_8) /
+                     (TEMP_RPI_THRESH_HIGH_C - TEMP_RPI_THRESH_LOW_C);
+
+  // Take the max percentage requested.
+  int16_t max_pcnt1   = amp_pcnt > psu_pcnt ? amp_pcnt : psu_pcnt;
+  int16_t max_pcnt_f8 = max_pcnt1 > rpi_pcnt ? max_pcnt1 : rpi_pcnt;
+  return max_pcnt_f8;
+}
+
+/* Calculates the fan duty cycle based on the desired fan percent
+ *
+ * Inputs
+ *    pcnt_f8: Desired fan percent in Q7.8 fixed-point format.
+ *
+ * Returns the duty cycle in UQ1.7 fixed-point format
+ */
+uint8_t dutyFromPercent(int16_t pcnt_f8) {
+  const int32_t scale_f16 = (1.0 - 0.3) * (1 << 16);
+  const int32_t min_f24   = 0.3 * (1 << 24);
+  int32_t       duty_f24  = scale_f16 * pcnt_f8 + min_f24;
+  return (uint8_t)(duty_f24 >> (24 - 7));
+}
+
+/* Calculates the digital potentiometer value to use to achieve the desired
+ * voltage. The dpot is nominally 10k Ohms.
+ *
+ * Inputs
+ *    pcnt_f8: Desired fan percent in Q7.8 fixed-point format.
+ *
+ * Returns the dpot value in the range [0x00,0x7F]
+ */
+uint8_t dpotValFromPercent(int16_t pcnt_f8) {
+  // Fans partially on, convert [0.0,1.0] in Q7.8 to [0x00,0x7F] dpot val
+  // Rpot (kOhms) = 10k * DPOT_VAL / 127 + 0.1
+  // V = 100k / (Rpot + 9k) + 1
+  //   = 100k / (10k * DPOT_VAL / 127 + 0.1 + 9k) + 1
+  //   = 100k / (10k / 127 * DPOT_VAL + 9.1k) + 1
+  // 10k / 127 * DPOT_VAL + 9.1k = 100k / (V - 1)
+  // DPOT_VAL = 127 / 10k * [100k / (V - 1) - 9.1k]
+  // DPOT_VAL = 1270 / (V - 1) - 115.57
+  const int32_t v_scale_f8  = (100 / 9.1 - 100 / 19.1) * (1 << 8);
+  const int32_t v_min_f16   = (100 / 19.1 + 1) * (1 << 16);
+  int32_t       v_fan_f16   = v_scale_f8 * pcnt_f8 + v_min_f16;
+  const int32_t numer_f20   = 1270 * (1 << 20);
+  const int32_t offset_f4   = 115.57 * (1 << 4);
+  int32_t       dpot_val_f4 = numer_f20 / (v_fan_f16 - (1 << 16)) - offset_f4;
+  return (uint8_t)(dpot_val_f4 >> 4);
+}
+
+/* Updates the fan state based on the current temp
+ *
+ * Inputs
+ *    amp_temp: Temperature of the amplifier heatsinks
+ *    psu_temp: Temperature of the high-voltage PSU
+ *    pi_temp:  Temperature of the Raspberry Pi
+ *    force:    Force fans on 100%
+ *    linear:   Digital potentiometer for linear voltage control is available
+ * All temps are in Q7.8 fixed-point format.
+ *
+ * Returns the current fan state.
+ */
+FanState* updateFans(int16_t amp_temp, int16_t psu_temp, int16_t rpi_temp,
```

Comment:
It looks like we can dynamically switch fan control methods since the control method detection is coupled with the control. Does this make sense? Should detection be separate and/or only done at startu

Reply:
tldr just make a comment explaining its dynamic

Reply:
Good point, I'm going to get this one in the next PR though since I've already touched a bunch of code removing the `state` variable.